include/seqan3/range/view/kmer_hash.hpp

```diff
@@ -12,47 +12,639 @@
+     *
+     * Constant.
+     *
+     */
+    auto size() const noexcept
+    //!\cond
+        requires std::SizedSentinel<std::ranges::sentinel_t<urng_t>, std::ranges::iterator_t<urng_t>>
+    //!\endcond
+    {
+        return end() - begin();
+    }
+};
+
+//!\brief A deduction guide for the view class template.
+template <std::ranges::ViewableRange orng_t>
+kmer_hash_view(orng_t &&, shape const shape_) -> kmer_hash_view<std::ranges::all_view<orng_t>>;
+
+// ---------------------------------------------------------------------------------------------------------------------
+// kmer_hash_fn (adaptor definition)
+// ---------------------------------------------------------------------------------------------------------------------
 
-    /*!\brief            Call the view's constructor with the underlying view as argument.
+//![adaptor_def]
+//!\brief view::kmer_hash's range adaptor object type (non-closure).
+struct kmer_hash_fn
+{
+    /*!\brief            Call the view's constructor with the underlying view and a seqan3::shape as argument.
      * \param[in] urange The input range to process. Must model std::ranges::ViewableRange and the reference type of the
      *                   range of the range must model seqan3::Semialphabet.
-     * \param[in] k      The k-mer size to construct hashes for.
+     * \param[in] shape_ The seqan3::shape to use for hashing.
      * \returns          A range of converted elements.
      */
-    template <std::ranges::ViewableRange urng_t>
-    //!\cond
-        requires Semialphabet<reference_t<urng_t>>
-    //!\endcond
-    constexpr auto operator()(urng_t && urange, size_t const k) const noexcept
+    template <std::ranges::Range urng_t>
+    constexpr auto operator()(urng_t && urange, shape const shape_) const
     {
-        return std::forward<urng_t>(urange) | ranges::view::sliding(k) | std::view::transform(
-        [] (auto const in)
-        {
-            std::hash<decltype(in)> h{};
-            return h(in);
-        });
+        static_assert(std::ranges::ViewableRange<urng_t>,
+            "The range parameter to view::kmer_hash cannot be a temporary of a non-view range.");
+        static_assert(std::ranges::ForwardRange<urng_t>,
+            "The range parameter to view::kmer_hash must model std::ranges::ForwardRange.");
+        static_assert(Semialphabet<reference_t<urng_t>>,
+            "The range parameter to view::kmer_hash must be over elements of seqan3::Semialphabet.");
+        return kmer_hash_view{std::forward<urng_t>(urange), shape_};
+    }
+    
+    //!\brief Store the shape and return a range adaptor closure object.
+    constexpr auto operator()(shape const shape_) const
+    {
+        return adaptor_from_functor{*this, shape_};
+    }
+    
+    //!\brief Construct a ungapped shape from the argument and return a range adaptor closure object.
+    constexpr auto operator()(ungapped const k) const
+    {
+        return adaptor_from_functor{*this, shape{k}};
```

Comment:
If you have implicit constructors (like you do right now), you can just remove this overload and a shape will automatically be created when an ungapped or bin_literal value is passed.

Reply:
Having implicit single-argument constructors here is ok from my POV, because the strong types are created explicitly for this purpose.